### PR TITLE
Adamic Adar and Resource Allocation Link Prediction Algorithms

### PR DESCRIPTION
**Pull request by mneedham**

### CODE REVIEW

**Comment by mneedham** [comment]
@davidoliverSP2 this is a new category for the algorithms. These PR has the code for them as well as the docs - hence why so many files changed!

File: readme.adoc [hunk 1]
```diff
@@ -73,6 +73,12 @@ These algorithms help calculate the similarity of nodes:
 * <<algorithms-similarity-euclidean, Euclidean Distance>> (`algo.similarity.euclidean`)
 * <<algorithms-similarity-overlap, Overlap Similarity>> (`algo.similarity.overlap`)
 
+=== Link Prediction
+
+These algorithms can be used as part of Link Prediction solution:
+
+* <<algorithms-linkprediction-adamic-adar, Adamic Adar Similarity>> (`algo.similarity.adamicAdar`)
+
```
Comment (davidoliverSP2) [inline readme.adoc]: should the **Resource Allocation algorithm** feature here too?

File: doc/asciidoc/linkprediction-adamic-adar.adoc [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+The library contains a function to calculate similarity between two nodes.
+
+// end::explanation[]
+
+[[algorithms-linkprediction-adamic-adar-usecase]]
+== Use-cases - when to use the Adamic Adar Similarity algorithm
+
+// tag::use-case[]
+We can use the Adamic Adar Similarity algorithm to work out the similarity between two nodes.
```
Comment (davidoliverSP2) [inline doc/asciidoc/linkprediction-adamic-adar.adoc]: We need to be careful with the consistency of naming this algorithm. It appears to be titled **The Adamic Adar algorithm**, but here it is titled **the Adamic Adar Similarity algorithm**. Above, it mentions _Adamic Adar similarity_, with a lower case "s".

I would recommend that we remove the "Similarity" from this example, and elsewhere in the page refer to only **The Adamic Adar algorithm** when we refer to "the algorithm".

File: doc/asciidoc/linkprediction-resource-allocation.adoc [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+where `N(u)` is the set of nodes adjacent to `u`.
+
+A value of 0 indicates that two nodes are dissimilar, while higher values indicate nodes are more similar.
+
+The library contains a function to calculate similarity between two nodes.
+
+// end::explanation[]
+
+[[algorithms-linkprediction-resource-allocation-usecase]]
+== Use-cases - when to use the Resource Allocation Similarity algorithm
```
Comment (davidoliverSP2) [inline doc/asciidoc/linkprediction-resource-allocation.adoc]: This page has the same issue regarding consistency of name.

Here, we use **the Resource Allocation Similarity algorithm**, whereas the title is **the Resource Allocation algorithm**. I would suggest sticking to the titled version, unless there is a good reason not to, but consistency is best practice.

File: readme.adoc (hunk 1 above)
Comment (mneedham) [reply readme.adoc]: Yes! Also I have the name of the procedure wrong so I've updated that as well

File: doc/asciidoc/linkprediction-resource-allocation.adoc (hunk 3 above)
Comment (mneedham) [reply doc/asciidoc/linkprediction-resource-allocation.adoc]: I have updated this to remove 'Similarity' everywhere. I was originally going to include this algorithms in the 'similarity' category, but realised that the way they're used is completely different and it didn't make sense to have them there.

File: doc/asciidoc/linkprediction-adamic-adar.adoc (hunk 2 above)
Comment (mneedham) [reply doc/asciidoc/linkprediction-adamic-adar.adoc]: Have removed similarity for these as well

File: algo/src/main/java/org/neo4j/graphalgo/linkprediction/LinkPrediction.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        return sumInverse(vector);
+    }
+
+
+    public double sumInverse(@Name("vector") List<Number> vector) {
+        // https://arxiv.org/pdf/0901.0553.pdf
+        if (vector == null) return 0;
+
+        double score = 0.0;
+        for (int i = 0; i < vector.size(); i++) {
```
Comment (jexp) [inline algo/src/main/java/org/neo4j/graphalgo/linkprediction/LinkPrediction.java]: foreach

File: algo/src/main/java/org/neo4j/graphalgo/linkprediction/LinkPrediction.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+                                       @Name(value = "config", defaultValue = "{}") Map<String, Object> config) {
+        ProcedureConfiguration configuration = ProcedureConfiguration.create(config);
+        RelationshipType relationshipType = configuration.getRelationship();
+        Direction direction = configuration.getDirection(Direction.BOTH);
+
+        List<Number> vector = new ArrayList<>();
+        try(Transaction tx = api.beginTx()) {
+            Set<Node> neighbors = findPotentialNeighbors(node1, relationshipType, direction);
+            neighbors.removeIf(node -> noCommonNeighbors(node, relationshipType, direction, node2));
+            neighbors.forEach(neighbor -> vector.add(degree(relationshipType, direction, neighbor)));
```
Comment (jexp) [inline algo/src/main/java/org/neo4j/graphalgo/linkprediction/LinkPrediction.java]: mapToDoulbe( nb -> 1.0/ degre(nb) ).sum()

File: algo/src/main/java/org/neo4j/graphalgo/linkprediction/LinkPrediction.java [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    @Description("algo.linkprediction.resourceAllocation(node1:Node, node2:Node, {relationshipQuery:'relationshipName', direction:'BOTH'}) " +
+            "given two nodes, calculate Resource Allocation similarity")
+    public double resourceAllocationSimilarity(@Name("node1") Node node1, @Name("node2") Node node2,
+                                       @Name(value = "config", defaultValue = "{}") Map<String, Object> config) {
+        ProcedureConfiguration configuration = ProcedureConfiguration.create(config);
+        RelationshipType relationshipType = configuration.getRelationship();
+        Direction direction = configuration.getDirection(Direction.BOTH);
+
+        List<Number> vector = new ArrayList<>();
+        try(Transaction tx = api.beginTx()) {
```
Comment (jexp) [inline algo/src/main/java/org/neo4j/graphalgo/linkprediction/LinkPrediction.java]: why a new tx?

File: algo/src/main/java/org/neo4j/graphalgo/linkprediction/LinkPrediction.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+import org.neo4j.procedure.UserFunction;
+
+import java.util.*;
+
+public class LinkPrediction {
+
+    @Context
+    public GraphDatabaseAPI api;
+
+    @UserFunction("algo.linkprediction.adamicAdar")
```
Comment (jexp) [inline algo/src/main/java/org/neo4j/graphalgo/linkprediction/LinkPrediction.java]: do you plan this also as a proc on the algo-graph?
then we could use intersection etc.

Comment (mneedham) [reply algo/src/main/java/org/neo4j/graphalgo/linkprediction/LinkPrediction.java]: Wasn't planning to as this one and resource allocation are used to find closeness between pairs of nodes. I don't think you'd use them to do all pairs similarity like we do with the similarity algos (Pearson, Cosine, Jaccard, etc)

File: algo/src/main/java/org/neo4j/graphalgo/linkprediction/LinkPrediction.java (hunk 6 above)
Comment (mneedham) [reply algo/src/main/java/org/neo4j/graphalgo/linkprediction/LinkPrediction.java]: I always forget that functions have an implicit tx!

File: algo/src/main/java/org/neo4j/graphalgo/linkprediction/LinkPrediction.java (hunk 5 above)
Comment (mneedham) [reply algo/src/main/java/org/neo4j/graphalgo/linkprediction/LinkPrediction.java]: Nice. Have updated

**Pull request merged**